### ISSUE TITLE
Pythonnet 2.1 not identifying System.Object type

### ISSUE THREAD

**Issue by vmuriart** [pythonnet/pythonnet#203]
I'm not sure how the tests wouldn't have caught this bug.
I have a `.net` class that has the function `AddQuery` with the following method signatures.

```
AddQuery
    Void AddQuery(api.QueryType)...
    Void AddQuery(System.Object)...
```

where `QueryType` is another custom object type. 

On `2.0` and `2.1dev1` the function works as expected. 
- I can pass a `QueryType` object to `AddQuery` and it would call the first method. 
- Calling `AddQuery` with a `string` or `int` would use the second method and process them.

On `2.1.0` and the earliest archived copy in `appveyor` this behavior changed. 
- Passing `QueryType` object to `AddQuery` still works. 
- Passing `string` type to `AddQuery` **errors** with `No method matches given arguments`

I've tested on `Python 2.7`, `Python 3.5` both x64. 
Looks like Pythonnet is no longer identifying `string` `int` and other types to be of type `System.Object` as well.

**Comment by vmuriart** [comment]
Possibly related to #187 
Tested on windows 7 and 8.1

**Comment by vmuriart** [comment]
Tried build before and after #176, both still have this bug present.

**Comment by denfromufa** [comment]
@vmuriart 
try passing `System.Object("your_python_string")` or `System.Object(123)` for `string` or `int` types respectively.

this breaking change is because implicit casting can be dangerous:

https://github.com/pythonnet/pythonnet/issues/131

**Comment by vmuriart** [comment]
@denfromufa just tried your suggestion and it semi-works. It passes through `pythonnet` and passes the `Object-type string` to the library I'm using, and breaks there. The underlying function is expecting a `String-type string` and fails to identify the `Object-type string` as such.

**Comment by denfromufa** [comment]
did you try `System.String()` and `System.Int32()`? or maybe even `System.Object(System.String())`...

**Comment by vmuriart** [comment]
I did. those to work, but have to explicitly cast each type before sending it. 
I'm taking a look at the src; I think there might be a better solution for the original bug #131 that would both fix the bug and not break legacy code. Might even try to fix #90 finally :smile:

**Comment by denfromufa** [comment]
@vmuriart I'm think of context manager that would allow you to specify "unsafe" casting operations like before for overloaded methods, see discussion here (my 4th comment):

github.com/pythonnet/pythonnet/issues/174

**Comment by denfromufa** [comment]
@vmuriart I think for your case, the overloading resolution in pythonnet could be improved by checking isinstance("string",System.Object) which corresponds to `is` operator in c#, but this is still dangerous - what if both types are present in 2 different overloaded methods?

**Comment by vmuriart** [comment]
I see your point. I'm not familiar enough with the source code in relation to how pythonnet decides which  method to call. 
The way I was thinking of it is that each Python type has an underlying c-type associated to it which should give an indication how to properly cast it into its equivalent `.net` counterpart. That may or may not work as a solution here though

**Comment by denfromufa** [comment]
@vmuriart python and .net types cannot be directly mapped to c-types due to much more complicated type system. all conversions are manually taken care for basic types in pythonnet:

https://github.com/pythonnet/pythonnet/blob/master/src/runtime/converter.cs

**Comment by vmuriart** [comment]
The problem that I am having with it at the moment is that quite a few of the libraries have broken with `2.1` release. Not sure if in this case the benefit outweighs the loss of backward compatibility. 
@tonyroberts it seems pr #151 has introduced a few incompatibilities in the attempt to solve #131. Can #151 be reverted until a better fix is in place?

**Comment by denfromufa** [comment]
Can we just look at all types of incompatibilities that you experience with
new method overloading?

Then we should really think whether the wrong results from
method overloading are less important than possible exceptions
with incompatible types? It would be very hard to spot the former and very
easy to fix or workaround the latter.

**Comment by tonyroberts** [comment]
@vmuriart @denfromufa my preference would be to attempt to fix 2.1 to fix the backward compatibilities, rather than just roll back. If it's just adding a case so that System.Object matches everything then hopefully that wouldn't be a big change, would it? Or are there more problems than that?

**Comment by tonyroberts** [comment]
@vmuriart would it be easy for you to add a few unit tests to check what's currently breaking? I think it's better to show the build currently as broken, if the only reason it's not showing as broken now is just lack of test coverage! :)

**Comment by denfromufa** [comment]
@tonyroberts so match even python `object` on `System.Object`?

**Comment by tonyroberts** [comment]
@denfromufa it may be more robust to check for base classes and interfaces of the mapped Python type

**Comment by denfromufa** [comment]
@tonyroberts @vmuriart can you provide me with few examples which need to be fixed?

**Comment by vmuriart** [comment]
I'll write up a couple tests for the issue

**Comment by vmuriart** [comment]
@denfromufa try this test.
It's failing with the same mechanism I saw earlier. 
https://github.com/vmuriart/pythonnet/commit/7acf93faa653b3a1d99f8546ad69322296587c6b

**Comment by denfromufa** [comment]
what would you expect for this test on python side when calling this CLR
method?

**Comment by vmuriart** [comment]
Just verified that this used to [work on the 2.0 version] (https://travis-ci.org/vmuriart/pythonnet/jobs/124006627) with this commit https://github.com/vmuriart/pythonnet/compare/e5c4e749b0bc...f77695f312e9

The test is pretty simple, I don't particularly care about the output just that the behavior still works.

**Comment by denfromufa** [comment]
i just noticed the python code, sorry for confusion.

What if the overloaded method with System.String is added? Then your tests
become ambiguous, so how should pythonnet handle this?

**Comment by denfromufa** [comment]
@vmuriart your tests are not complete - you do not check that correct overloaded method is selected. Some sort of asserting is required.

**Comment by vmuriart** [comment]
Correct. I meant the test as a proof of concept to demonstrate how to repeat the bug and that the behavior worked on the old version.

**Comment by denfromufa** [comment]
@vmuriart sorry - I'm still not clear, what do you expect, if overloaded method with System.String argument is added? 

this is definitely not bug, since no agreed upon test is available for this scenario as of yet and implicit overloading is not supported anymore.

**Comment by tonyroberts** [comment]
@denfromufa if an overload with argument type System.String was added then that would be used instead when a string was passed. If a non string type (but one still deriving from System.Object, e.g. System.Int32) was used then the object method would be used.

There's no implicit conversion from one type to another going on, a string *is an* object so the overload is valid.

Here's a link that might be useful
https://msdn.microsoft.com/en-us/library/aa691356(v=vs.71).aspx

**Comment by denfromufa** [comment]
@tonyroberts thank you for this explanation. Let me now write this code for base classes.

**Comment by vmuriart** [comment]
@tonyroberts thanks for explaining it! I was struggling on getting the right words together.

**Comment by matthid** [comment]
Just an idea: Maybe Roslyn can help with issues like these. If we can re-use their (or parts of their) overload resolution logic we don't need to reimplement the spec.

On the other hand: Going a different route is valid as well. In F# you actually need to mark implicit casts with an operator (while this case would still work as everything is an object)

**Comment by denfromufa** [comment]
@matthid i was thinking about re-using roslyn as well, but then we have a lot of python-specific logic.

what do you mean by "On the other hand: Going a different route is valid as well"?

**Comment by matthid** [comment]
What I meant was that pythonnet doesn't need to mirror C# overload resolution logic. Instead it is completely fine to define its own (possibly simpler) logic. For this issue I'd say that most people expect this to work. But IMHO it could be closed as "wontfix" as well (and say that "by-design" types need to match exactly).

It depends on what spec/design guideline this project follows. As you are trying to mirror ironpython(?) this is most likely a bug :)

**Comment by denfromufa** [comment]
it appears that precedence of types is simplified version of Jython, found by @vmuriart 👍 

https://github.com/pythonnet/pythonnet/blob/master/src/runtime/methodbinder.cs#L170

I'm not sure how this compares to IronPython, but likely very similar due to the same author:
https://en.wikipedia.org/wiki/Jim_Hugunin

**Comment by vmuriart** [comment]
Fun fact on the history of Jim and IronPython.

> IronPython was created in 2005 by Jim Hugunin to prove that the .NET Framework was a poor platform for dynamic languages. He failed to do so, and IronPython was born.
